File: scripts/python/parsers/named_credential_parser.py
"""Parser for Named Credential metadata (.namedCredential-meta.xml)."""
import re
from pathlib import Path


def _extract_tag(xml: str, tag: str) -> str | None:
    m = re.search(rf'<{tag}[^>]*>([\s\S]*?)</{tag}>', xml)
    return m.group(1).strip() if m else None


def parse_named_credentials(type_dir: str) -> dict:
    """
    @param type_dir: path to the namedCredentials/ folder
    @returns: { credentials: list }
    """
    credentials = []
    p = Path(type_dir)
    if not p.exists():
        return {'credentials': credentials}

    for f in sorted(p.glob('*.namedCredential-meta.xml')):
        xml = f.read_text(encoding='utf-8')
        credentials.append({
            'apiName':       f.name.replace('.namedCredential-meta.xml', ''),
            'label':         _extract_tag(xml, 'label'),
            'endpoint':      _extract_tag(xml, 'endpoint'),
            'principalType': _extract_tag(xml, 'principalType'),
            'protocol':      _extract_tag(xml, 'protocol'),
        })

    return {'credentials': credentials}

File: scripts/node/parsers/namedCredentialParser.js
'use strict';

const fs   = require('fs');
const path = require('path');

function extractTag(xml, tag) {
  const m = xml.match(new RegExp(`<${tag}[^>]*>([\\s\\S]*?)<\\/${tag}>`));
  return m ? m[1].trim() : null;
}

/**
 * Reads Named Credential metadata from SFDX source format.
 * @param {string} typeDir - path to the namedCredentials/ folder
 * @returns {{ credentials: Array }}
 */
async function parseNamedCredentials(typeDir) {
  const credentials = [];
  if (!fs.existsSync(typeDir)) return { credentials };

  const files = fs.readdirSync(typeDir).filter(f => f.endsWith('.namedCredential-meta.xml'));

  for (const file of files) {
    const xml = fs.readFileSync(path.join(typeDir, file), 'utf8');
    credentials.push({
      apiName:        file.replace('.namedCredential-meta.xml', ''),
      label:          extractTag(xml, 'label'),
      endpoint:       extractTag(xml, 'endpoint'),
      principalType:  extractTag(xml, 'principalType'),
      protocol:       extractTag(xml, 'protocol'),
    });
  }

  return { credentials };
}

module.exports = { parseNamedCredentials };
